### ISSUE TITLE
Easily use named file descriptors passed by systemd

### ISSUE THREAD

**Issue by quasicomputational** [haproxy/haproxy#766]
## What should haproxy do differently? Which functionality do you think we should add?

For HAProxy to learn to call `sd_listen_fds_with_names` itself, with a syntax like `bind "named-fd@foo"` meaning "bind to the passed file descriptor named `foo`".

## What are you trying to do?

Start (& restart) HAProxy with privs already dropped. This entails HAProxy listening on sockets passed as file descriptors, as binding can be a privileged operation.

## Context

systemd has the ability to pass more than one file descriptor to a service. It does not make any guarantees about which order the file descriptors are passed in. Instead, the sockets are given names in their unit files. If services want more than one listening socket, they must call [`sd_listen_fds_with_names`](https://www.freedesktop.org/software/systemd/man/sd_listen_fds.html) to learn which file descriptor corresponds to which name.

However, currently, HAProxy only supports using passed file descriptors with either constant numbers or numbers passed a environment variables using the `bind "fd@3"` or `bind "fd@{ENV_VAR}"` syntaxes. Hence, because file descriptors may be passed in arbitrary order, this cannot be used directly if HAProxy needs to listen on more than one file descriptor (because, e.g., of the need to be bound to specific IPs or to listen on different network interfaces).

While it's possible to achieve this today by invoking HAProxy through a wrapper that'll call `sd_listen_fds_with_names`, unset close-on-exec, and pass the file descriptor numbers it learns in environment variables, it'd be handy for HAProxy to learn about this directly and avoid the dance.

**Comment by wtarreau** [comment]
This could indeed be a good idea.

However I tend to be careful with passed FDs. The risk of doing something wrong with them that the owner doesn't notice is quite high. For example what happens if you call "shutdown frontend foo" on the CLI and it shuts the listening socket down (or the FD is passed by accident to an external program doing it)? If systemd doesn't detect this, it will continue to pass this socket to future invocations of the program, and no more traffic will be received until a reboot (or some other ugly maintenance operation that should never happen with a regular init).

But what I'm saying here remains true in general with fd passing, except that usually it's easy to stop/start a service while here I'm not sure the user keeps control over this. At the very least it should be documented so that users weigh the pros and cons before blindly engaging down that route.

**Comment by TimWolla** [comment]
> For example what happens if you call "shutdown frontend foo" on the CLI and it shuts the listening socket down (or the FD is passed by accident to an external program doing it)?

The [systemd documentation says](https://www.freedesktop.org/software/systemd/man/systemd.socket.html#Accept=):

> It should not invoke shutdown(2) on sockets it got with Accept=no

**Comment by wtarreau** [comment]
Yes, thus definitely one more reason for placing a big warning about this in the doc!

**Comment by wlallemand** [comment]
I have an old patch which does this https://github.com/haproxy/haproxy/commit/61937cf2e1138410f97dccc85e70594781d1ac23 

But I wonder if it won't be better to support the syntax "sd@SERVICENAME" instead.

**Comment by sepointon** [comment]
> While it's possible to achieve this today by invoking HAProxy through a wrapper that'll call sd_listen_fds_with_names, unset close-on-exec, and pass the file descriptor numbers it learns in environment variables, it'd be handy for HAProxy to learn about this directly and avoid the dance.

This actually won't work quite right, because HAProxy won't set cloexec on the FDs, and the wrapper obviously can't, so there's some FD leak hazard. So, this really does need some smarts in HAProxy somewhere.
